opencxl/apps/cxl_image_classification_host.py

```diff
@@ -103,30 +112,100 @@ async def write_mmio(self, address: int, size: int, value: int):
+        self._validation_results: List[List[Dict[str, float]]] = [[] for _ in self._total_samples]
+        pic_id = 0
+        pic_data_mem_loc = 0x00008000
+        for c in categories:
+            category_pics = glob.glob(f"{c}/*.JPEG")
+            sample_pics = sample(category_pics, self._sample_from_each_category)
+            category_name = c.split(os.path.sep)[-1]
+            self._sampled_file_categories += [category_name] * self._sample_from_each_category
+            for s in sample_pics:
+                with open(s, "rb") as f:
+                    pic_data = f.read()
+                    pic_data_int = int.from_bytes(pic_data, "little")
+                    pic_data_len = len(pic_data)
+                    await self.store(pic_data_mem_loc, pic_data_len, pic_data_int)
+                    for dev in range(self._device_count):
+                        # Remember to configure the device when starting the app
+                        # Should make sure to_device_addr returns correct mmio for that dev
+                        # In fact, we can use a fixed host memory addr
+                        # and we only need to write the length to the device
+                        await self.write_mmio(
+                            self.to_device_mmio_addr(dev, 0x810), 8, pic_data_mem_loc
+                        )
+                        await self.write_mmio(self.to_device_mmio_addr(dev, 0x818), 8, pic_data_len)
+                        event = asyncio.Event()
+                        self._irq_handler.register_interrupt_handler(
+                            Irq.ACCEL_VALIDATION_FINISHED,
+                            self._save_validation_result_type1(dev, pic_id, event),
+                        )
+                        await self._irq_handler.send_irq_request(Irq.HOST_SENT, dev)
+                        await event.wait()
+                        # Currently we don't send the picture information
+                        # (e.g., pic_id) to the device
+                        # and to prevent race condition, we need to send pics synchronously
+                    pic_data_mem_loc += pic_data_len
+                    pic_id += 1
+        self._merge_validation_results()
+
+    def _save_validation_result_type1(self, dev: int, pic_id: int, event: asyncio.Event):
+        async def _func():
+            # We can use a fixed host_result_addr, say 0x0A000000
+            # Only length is needed
+            host_result_addr = await self.read_mmio(self.to_device_mmio_addr(dev, 0x820), 8)
+            host_result_len = await self.load(self.to_device_mmio_addr(dev, 0x828), 8)
+            data = await self.load(host_result_addr, host_result_len)
+            data_bytes = data.data.to_bytes(host_result_len, "little")
+            validate_result = json.loads(data_bytes)
+            self._validation_results[pic_id].append(validate_result)
+            event.set()
+
+        return _func
```

Comment:
Given,
```
    def register_interrupt_handler(self, irq_msg: Irq, irq_recv_cb: Callable):
        async def _callback():
            await ev.wait()
            irq_recv_cb()
```
`_save_validation_result_type1()` will be called without `await`, even though it returns an `async` function. Try utilizing `create_task()` and `gather()` instead.

Reply:
In fact, all functions being called by invoking irq_recv_cb are async functions. Changed
`irq_recv_cb()` to `await irq_recv_cb()`.